File: android/vbt-app/app/src/main/java/com/vbt/app/ui/screen/plans/PlanListViewModel.kt
package com.vbt.app.ui.screen.plans

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.vbt.app.data.remote.ApiService
import com.vbt.app.data.remote.TrainingPlanDto
import com.vbt.app.data.remote.UserDto
import com.vbt.app.data.repository.AuthRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class PlanListUiState(
    val plans: List<TrainingPlanDto> = emptyList(),
    val athletes: List<UserDto> = emptyList(),
    val isCoach: Boolean = false,
    val isLoading: Boolean = false,
    val error: String? = null
)

@HiltViewModel
class PlanListViewModel @Inject constructor(
    private val apiService: ApiService,
    private val authRepository: AuthRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(PlanListUiState())
    val uiState: StateFlow<PlanListUiState> = _uiState.asStateFlow()

    init {
        loadPlans()
    }

    fun loadPlans() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)
            try {
                authRepository.getCurrentRole().collect { role ->
                    val isCoach = role == "coach"
                    _uiState.value = _uiState.value.copy(isCoach = isCoach)

                    val plansResponse = apiService.getPlans()
                    if (plansResponse.isSuccessful) {
                        val plans = plansResponse.body() ?: emptyList()

                        if (isCoach) {
                            val athletesResponse = apiService.getAthletes()
                            val athletes = if (athletesResponse.isSuccessful) {
                                athletesResponse.body() ?: emptyList()
                            } else {
                                emptyList()
                            }
                            _uiState.value = _uiState.value.copy(
                                plans = plans,
                                athletes = athletes,
                                isLoading = false
                            )
                        } else {
                            _uiState.value = _uiState.value.copy(
                                plans = plans,
                                isLoading = false
                            )
                        }
                    } else {
                        _uiState.value = _uiState.value.copy(
                            error = "Failed to load plans: ${plansResponse.code()}",
                            isLoading = false
                        )
                    }
                }
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    error = "Error: ${e.message}",
                    isLoading = false
                )
            }
        }
    }

    fun deletePlan(planId: Int) {
        viewModelScope.launch {
            try {
                val response = apiService.deletePlan(planId)
                if (response.isSuccessful) {
                    loadPlans()
                } else {
                    _uiState.value = _uiState.value.copy(
                        error = "Failed to delete plan: ${response.code()}"
                    )
                }
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    error = "Error deleting plan: ${e.message}"
                )
            }
        }
    }

    fun assignPlan(planId: Int, athleteId: Int) {
        viewModelScope.launch {
            try {
                val response = apiService.assignPlanToAthlete(planId, athleteId)
                if (response.isSuccessful) {
                    loadPlans()
                } else {
                    _uiState.value = _uiState.value.copy(
                        error = "Failed to assign plan: ${response.code()}"
                    )
                }
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    error = "Error assigning plan: ${e.message}"
                )
            }
        }
    }
}

File: ios/VBT/VBT/Features/Plans/PlanListViewModel.swift
import Foundation
import Observation

/// Port 1:1 z Android `PlanListViewModel.kt`.
@Observable
final class PlanListViewModel {
    private(set) var plans: [TrainingPlanDto] = []
    private(set) var athletes: [UserDto] = []
    private(set) var isCoach = false
    private(set) var isLoading = false
    var error: String?

    private let apiClient: APIClient
    private let authRepository: AuthRepository

    init(apiClient: APIClient, authRepository: AuthRepository) {
        self.apiClient = apiClient
        self.authRepository = authRepository
        self.isCoach = authRepository.currentRole == .coach
    }

    @MainActor
    func loadPlans() async {
        isLoading = true
        error = nil
        do {
            plans = try await apiClient.send(.getPlans())
            if isCoach {
                athletes = (try? await apiClient.send(.getAthletes())) ?? []
            }
        } catch {
            self.error = "Nie udało się wczytać planów: \(error.localizedDescription)"
        }
        isLoading = false
    }

    @MainActor
    func deletePlan(_ planId: Int) async {
        do {
            try await apiClient.sendNoContent(.deletePlan(id: planId))
            await loadPlans()
        } catch {
            self.error = "Nie udało się usunąć planu: \(error.localizedDescription)"
        }
    }

    @MainActor
    func assignPlan(_ planId: Int, to athleteId: Int) async {
        do {
            let _: TrainingPlanDto = try await apiClient.send(.assignPlan(id: planId, athleteId: athleteId))
            await loadPlans()
        } catch {
            self.error = "Nie udało się przypisać planu: \(error.localizedDescription)"
        }
    }
}
